### PR TITLE
Fix an issue with types possibly being related to discriminated type despite of incompatible variance

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/51180

This union was correctly failing a check here:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L19750
but its result isn't directly used if the check failed. This check effectively went through `isRelatedTo` to this block and failed there:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L20150-L20171

So since this correct result of `Ternary.False` was not returned there (for reasons I don't understand but they seem to be important), the execution has continued. It continued up to this check here which succeeded:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L20219

In there, only the structures are compared here:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L20370-L20383

I've tried to call `isRelated` there instead but that experiment failed (maybe somebody else could make it work though). So I've reused the code related to variances that was responsible for creating the "initial" correct result of `Ternary.False`.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20367,6 +20367,28 @@ namespace ts {
...
+                        // We return Ternary.Maybe for a recursive invocation of getVariances (signalled by emptyArray). This
+                        // effectively means we measure variance only from type parameter occurrences that aren't nested in
+                        // recursive instantiations of the generic type.
+                        if (variances === emptyArray) {
+                            return Ternary.Unknown;
+                        }
+                        if (!typeArgumentsRelatedTo(getTypeArguments(source as TypeReference), getTypeArguments(type as TypeReference), variances, /* reportErrors */ false, IntersectionState.None)) {
+                            return Ternary.False;
+                        }
+                    }
```
Comment (Andarist) [inline src/compiler/checker.ts:20390]: This is effectively a copy-paste (with very small adjustments) of the block here:
https://github.dev/microsoft/TypeScript/blob/4c9afe8812fcdb4658472ccbced4a5cd4bae70ea/src/compiler/checker.ts#L20150-L20171

This definitely should be deduplicated if the solution itself is correct. I will wait for the review before making any changes to this.

**Comment by Andarist** [comment]
It's kinda wild that the CI fails here with a type error. I can't repro this locally (using the same script) and I don't know how I would get a type error at that location as the changes in this PR are totally unrelated to it.

**Comment by RyanCavanaugh** [comment]
The failure here is saying that this change causes TS to no longer be able to build itself. Confirming, I see this locally checking out your PR branch, but not in main.

You can run
```
gulp local --built
```
to reproduce

**Comment by RyanCavanaugh** [comment]
Aside, it's quite rare to break TS without affecting any other tests! We'd want a testcase that isolates the problem this exposes, assuming there is one.

**Comment by Andarist** [comment]
Thanks for the tip on how to reproduce this locally! I was trying `gulp local` before (guided by the error logs) but I didn't use the `--built` flag (nor `--lkg=false` that is used by the CI).

I've checked out the failing call site and as far as I can tell - it's a legit one :P it shouldn't fail. I've created a TS playground demonstrating the problem: [TS playground](https://www.typescriptlang.org/play?ts=4.9.0-dev.20221018#code/JYWwDg9gTgLgBAbzgZQJ4DsYEMAeBpYdAEwBo4BZCI4AM2AFMoAVCAa3vTMursbgF84NKBBBwARDFRh6AZwDGUYGBjiA3HAD0muFJmyAXHAAsAOgCcpgAwBaIvQBupgExXnzgIxWPADgBQfnr0cABK9DAArlDo9EQAojhgUHKywBDoTNLBALwUVLQMzGwcADx+cHAAPigY2PiERKYAwumyMHj0qADu0ETlVTWYuATEpgAi9DRYEQA27Z09UH0V1WhD9aMJkLAd3b39q7XDDaYAkui7i8sDa3UjjQAKSg5YMPSX+yuDdydPEG-yN5ED5LA7fY6jB4RABGM2A8hB10O63upmQ2Bg8MRYNuEMaAEFoW0oFhAdivriNgT5PIUtByTcjlTTPjZBgEQtPoyUScJvIZlhkgzkT9RgB5CLzPagilM1FhLBEdIzVDC8HMsUORhKezYgB8agCQTg+PQ6H+r1iYUi0UyMjguW4BUYhr89n5guC8la8HoiWYWSM1qiMXi-pSaQyWUN2jg8iiyUwKt0AAtgLI4Om4FgZjMIF1YnBHBxUxAIgBzFOp4LYKDl8K6LKZjNYOARdCRuBdFOMehYDMwHtwWRlqC0xv2lP97NtjvpOAvGYRYLzwfBAAGrAa67gSQgMlgqD8scHWdk9HoIAHEDg0OCYrww8I4-XTt4RXY6B3Srk6AA5PAyQ5smU5atm6CoHA-w9lAu4iAemJyKmrxdmWMxELewQjiA9Apvms7AAAjsuug3q2sgyPIBTyBO9DHjowAwH+GZ7tCWCwpB3aQWu2ZEPYGEvEoWDoOO8g9vIrBwD+sj-vAwH4aeLayKk5boDhmDgRhPHEhEgJRDmcaiGAgrpvOWbmvAETnhhtDVguJnCaJ4mSVMwAzLIfh+ts8DeugbRwMkMB2vQRimhZlpEMGtpNrkXkBjIrrGgAYsAOBWuEIaxFsyTKZGwUOmCb6FCwn4lJSqItH5UpXHqhX5O+JWlOVvKTNMcyIrVXxFYwjXoGVconFs0DVb0nUDN1H5NQNoznB1dU8MVxR9c1kLPJac1dfVi2lStjwiACQIbeNW09Ut-U8pCMJwhy0pEGN1QTb152io06KvFinJLPdeQLadO3TQSRIwCSZKfXd83OpNy0AyyNJ0lAR0PSdUPPXiLJsiJiM-ZDT27eM9AekKYPfY9Z14xKI1fRDDVkzDCpKugKpY6T-0XY0mrasAurE667oCskhlVUW-rBc4RgpWlkUZdEWXhrl6TBTGOiKYZ4BuUhD63pKyFMRmW7EK2wBQMkMyOMJ8CpCJNZDtBfCnug5bDnhswYXe2a5ap7Gm6RECeYkw2C-5gWi6FZoWkCUVRvasUi1kziGkAA). Gonna wrap this up later in a test case that fails with this PR but passes with the main branch and gonna open a PR with that directly to main.

In the meantime, maybe you'd have suggestions on how this PR could be improved to fix the original issue without degrading this case?

**Comment by Andarist** [comment]
OTOH - IIRC variance check **can** return a different result than a full structural check and that's intended... 🤔

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts:20390]: Using variance results more is usually good for perf, but a variance-based result and a structural result being different is fundamentally the problem. Usually means some comparison needs to add an `Unmeasurrable` or `Unreliable` marker to the result calculated for the variance relationship.

Comment (Andarist) [reply src/compiler/checker.ts:20390]: Do you think there is a problem with the added code? Or that there is a problem elsewhere and that elsewhere Unmeasurable/Unreliable is not handled correctly?

**Comment by weswigham** [comment]
> IRC variance check can return a different result than a full structural check and that's intended... 🤔

Definitely not(*). Any time the variance measurement for a type differs from the structural check, it's a bug, usually in the variance calculation logic. (We have many structural type relationships that aren't strictly co- or contra- variant and so have `Unreliable` and `Unmeasurable` bail-outs in variance measurement to flag those cases so we rely partially or solely on structural results for them, respectively).

* Unless `in` and `out` annotations are used, then it's anyone's guess. We toss a lot of variance measurement correctness guarantees out the window when those get used, since they can often be more restrictive than the measured variance, but only get used for variance-based comparisons, and not structural ones. The resulting mismatch can cause a headache of weird behavior which I usually just warn people off with "only use `in` and `out` for performance, and only if you know what you're doing".

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts:20390]: More likely that a non-strictly-co--or-contra-variant comparison somewhere needs to throw out a `Unmeasurable` or `Unreliable` flag but isn't yet doing so.

**Review by weswigham — CHANGES_REQUESTED** [review]
Just formally marking changes requested - based on the behavior, there's definitely a comparison _somewhere_ whose variance result needs to be marked `Unreliable` and today is not. Tracking that down and marking it is the correct fix.

**Comment by Andarist** [comment]
This has fallen off my radar and pushed down the priority list but I will be back here to investigate this further (unless somebody beats me to it). For now, I will convert this to a draft to indicate that I don't exactly work on this *right* now.

**Comment by Andarist** [comment]
@weswigham going to close this PR since you have fixed the inconsistency in https://github.com/microsoft/TypeScript/pull/52106 but part of the issue still looks suspicious and perhaps could be looked into or explained (see [here](https://github.com/microsoft/TypeScript/issues/51180#issuecomment-1576709965))

**Pull request closed without merging**